File: process/R/dmm.R

run_dmm <- function(population_data, cfg, start_year, years = 5) {
  print("Starting DMM Processing")
  
  results <- list()
  end_year <- start_year + years - 1
  
  results_index <- 1
  
  start_pop <- population_data
  start_pop$year = start_pop$base_year
  
  for (proc_year in start_year:end_year) {
    print(paste0("Processing Year: ", proc_year))
    
    proc_pop <- person_env$forward(start_pop, proc_year, cfg)
    proc_pop <- mortality_env$forward(proc_pop, cfg)
    # fertility forward (TBA)
    proc_pop = employment_env$forward(proc_pop, cfg)
    
    print(paste0(
      proc_year, ": ", 
      mean(proc_pop$latent_market_income), 
      "/", 
      nrow(proc_pop[proc_pop$life_stage == "dead",])))
    results[[results_index]] <- proc_pop
    
    results_index <- results_index + 1
    
    start_pop = proc_pop
  }
  
  results <- bind_rows(results)
  
  output_path <- paste0(cfg[["output_dirs"]][["outputs"]], "/results.parquet")
  
  print(paste0("Writing outputs: ", output_path))
  
  dir.create(cfg[["output_dirs"]][["outputs"]], recursive = TRUE, showWarnings = FALSE)
  
  write_parquet(as.data.frame(results), output_path)
  
}

File: process/Python/dmm.py
from logging import getLogger
from process.Python.person import forward as person_forward
from os.path import exists
from os import makedirs
from pandas import concat
from pyarrow.parquet import write_table as pq_write_table
import pyarrow as pa

logger = getLogger(__name__)


def run_dmm(population_data: dict, cfg: dict, start_year: int, years: int = 5):

    logger.info("Starting DMM Processing")

    results = []
    for proc_year in range(start_year, start_year + years):

        logger.info(f"Processing Year: {proc_year}")

        proc_pop = person_forward(population_data, proc_year, cfg)

        results.append(proc_pop)

    results = concat(results)

    results = results.reset_index()

    output_path = f"{cfg["output_dirs"]["outputs"]}/mortality_data.parquet"

    print(f"Writing outputs: {output_path}")

    if not exists(cfg["output_dirs"]["outputs"]):
        makedirs(cfg["output_dirs"]["outputs"])

    pq_write_table(pa.Table.from_pandas(results), output_path)
